### PR TITLE
Simple Mapping type improvements

### PR DESCRIPTION
**Pull request by Robbepop**
Based on @HCastano 's https://github.com/paritytech/ink/pull/946 PR.

This fork of the above PR improves the `Mapping` type even further which reduces the ERC-20 example smart contract Wasm file size to 10.8kB so far.

# Implementation Notes

- The ink! codegen does not automatically generate `SpreadAllocate` implementations of its `#[ink(storage)]` struct and at this point it is even undecided if that's a good idea. Therefore we currently are required to implement `SpreadAllocate` manually in the ERC-20 example contract.
- The ERC-20 `new` constructor as well as the `new_init` method seem a bit verbose but we are working on some improvements to make this less of a deal.
- The `EncodeLike` trait usage in the `Mapping` type allows us to use `&(&K1, &K2)` in the `Mapping::{get, insert}` methods instead of just `&(K1, K2)` which saves us a lot of Wasm file size overhead with respect to big value types such as `Balance` and especially `AccountId`.
- To avoid dealing with big value types helper methods have been implemented in the ERC-20 example contract such as `allowance_impl` and `balance_of_impl` that take and forward their arguments by reference instead of by value.

### CODE REVIEW

File: crates/storage/src/collections/mapping/mod.rs
Comment (xgreenx) [reply crates/storage/src/collections/mapping/mod.rs:94]: I think we don't need to specify the `inline` attribute. Because the compiler will decide it based on the optimization flag(in our case it reduces the size of the code). So compiler better knows what better to inline to reduce the size of the contract=)
https://nnethercote.github.io/perf-book/inlining.html

Comment (HCastano) [reply crates/storage/src/collections/mapping/mod.rs:94]: Just tried now, and there are no changes to the Wasm size. However, there are also no changes if I remove the `inline`'s from `insert()` and `get()`, so maybe we should get rid of those as well?

Comment (athei) [reply crates/storage/src/collections/mapping/mod.rs:94]: Not sure if the compiler also takes into account that passing large types (> 64bit) is very expensive in wasm. Ideally, `Q` would be a reference.

File: crates/storage/src/collections/mapping/mod.rs
Comment (HCastano) [inline crates/storage/src/collections/mapping/mod.rs:74]: Can I think of `EncodeLike` sort-of like `EncodeLike = Borrow + Encode`?

File: examples/erc20/lib.rs
Comment (HCastano) [inline examples/erc20/lib.rs:63]: Should we add a note here that this is intended to only be temporary and will (hopefully) be replaced by #995?

File: examples/erc20/lib.rs
Comment (HCastano) [reply examples/erc20/lib.rs:141]: After reading the code more carefully I get why 👍

**Review by HCastano — APPROVED** [review]

File: crates/storage/src/collections/mapping/mod.rs
Comment (Robbepop) [reply crates/storage/src/collections/mapping/mod.rs:74]: EncodeLike is a purely marker trait coming from the parity-scale-codec crate. Please go to the crates docs to read more about its general use.

File: examples/erc20/lib.rs
Comment (Robbepop) [reply examples/erc20/lib.rs:63]: I think it is better to merge this PR after merging the one PR that improves the SpreadLayout traits.

File: crates/storage/src/collections/mapping/mod.rs
Comment (HCastano) [reply crates/storage/src/collections/mapping/mod.rs:74]: I read the docs, and I think my question is still valid 😄

I think the semantics of a type being able to be SCALE encoded as another type (`EncodeLike`) vs. being able to be borrowed as another type (`Borrow`) are slightly different, which is why I asked

**Comment by forgetso** [comment]
Is there a plan for a function like `get_mut` as part of `Mapping`? How do we update a key value pair - just insert again?

**Comment by xgreenx** [comment]
> Is there a plan for a function like `get_mut` as part of `Mapping`? How do we update a key value pair - just insert again?

Yes, insert again=)

**Comment by Robbepop** [comment]
> Is there a plan for a function like `get_mut` as part of `Mapping`? How do we update a key value pair - just insert again?

A `get_mut` function won't be part of the simple `Mapping` type but it will be available on types that build upon the `Mapping` type internally and provide a more high-level API.

**Comment by forgetso** [comment]
Thanks for the info. Is `iter` deliberately not implemented for `Mapping`? I started converting my contract to `Mapping` but I have methods that previously used `hashmap.keys()`. Now I can't seem to create an iterator for `Mapping` to implement  a `keys` style function.
```
    impl ProviderMapping {
        fn keys(&self) -> Vec<AccountId> {
            let mut keys: Vec<AccountId> = Vec::new();
            for (key, value) in &self.providers.iter() {
                keys.push(key);
            }
            keys
        }
```



```
for (key, value) in &self.providers.iter() {
   |                                                 ^^^^ method not found in `Mapping<ink_env::AccountId, Provider>`
```

**Comment by Robbepop** [comment]
If you come from the Solidity world then `Mapping` is literally just what Solidity provides to you with its own builtin mapping types which are ... just mappings and no real collections that do have knowledge about their contained elements.
So it is not the job of the `Mapping` type to provide collection APIs such as `iter`. This is a very common misunderstanding unfortunately.
If what you need is a collection then you can either build one yourself using the `Mapping` type internally OR use ink!'s `ink_storage::HashMap`.

**Comment by forgetso** [comment]
Cool this is my first time coding smart contracts and I'm coming from a python/JS background, never written in Rust/C. I'll try creating an iterable `Mapping` as per [this example](https://solidity-by-example.org/app/iterable-mapping/) and compare contract size to `Hashmap`. Thanks for the clarification, you probably saved me a few more hours of days of searching.

**Comment by xgreenx** [comment]
> Cool this is my first time coding smart contracts and I'm coming from a python/JS background, never written in Rust/C. I'll try creating an iterable `Mapping` as per [this example](https://solidity-by-example.org/app/iterable-mapping/) and compare contract size to `Hashmap`. Thanks for the clarification, you probably saved me a few more hours or days of searching.

Better to avoid iteration through the map in the smart contracts because it can contain a lot of entries. If you need to resolve some local issue(for example, in your algorithm, you need properties of the `map` to solve some issues. But it is needed only during function execution and after that, you don't need to store that `map` into storage), you can use `ink_prelude::collections::HashMap`. Or if you know that count of entries is not big, then you can store `ink_prelude::collections::HashMap` into the storage(when you will load/save that map you will load/save the whole map with all entries).

**Comment by xgreenx** [comment]
> I have reworked my contract using `Mapping` instead of `Hashmap`, which has reduced the size from `59.8K` to `45K` when building on Supercolony's [reworked-common-part-of-message](https://github.com/Supercolony-net/ink/tree/optimizations/reworked-common-part-of-message) branch. However, I also removed any get functions and replaced them with `Vectors` that contain known keys for `Mappings` - so that will have helped.
> 
> One of the issues I ran into during this process is that `enum` is not supported by `SpreadAllocate`, which seems to be [required as part of the new contract instantiation process](https://github.com/paritytech/ink/pull/995#discussion_r755031713).
> 
> I had to [modify](https://github.com/prosopo-io/ink/commit/bdc676d61804a3fd88f76cb8045238250b3db166) `ink` to allow `enums` to implement `SpreadAllocate`. Hopefully this will be allowed in `ink` in future.
> 
> I also tried and failed to generalise my types that build on `Mapping`, as mentioned [here](https://github.com/paritytech/ink/pull/995#discussion_r756155984).

I tried your contract with [a new version](https://github.com/paritytech/ink/pull/1017) and it takes `Original wasm size: 81.7K, Optimized: 32.4K(32355)`.

**Pull request merged**
